Firestore/core/src/firebase/firestore/util/error_apple.h

```diff
@@ -38,48 +43,45 @@ namespace firestore {
 namespace util {
 
 // Translates a set of error_code and error_msg to an NSError.
-inline NSError* MakeNSError(const int64_t error_code,
-                            const absl::string_view error_msg,
-                            NSError* cause = nil) {
-  if (error_code == FirestoreErrorCode::Ok) {
-    return nil;
-  }
-
-  NSMutableDictionary<NSString*, id>* user_info =
-      [NSMutableDictionary dictionary];
-  user_info[NSLocalizedDescriptionKey] = WrapNSString(error_msg);
-  if (cause) {
-    user_info[NSUnderlyingErrorKey] = cause;
-  }
-
-  return [NSError errorWithDomain:FIRFirestoreErrorDomain
-                             code:static_cast<NSInteger>(error_code)
-                         userInfo:user_info];
-}
+NSError* MakeNSError(const int64_t error_code,
+                     const absl::string_view error_msg,
+                     NSError* cause = nil);
 
 inline NSError* MakeNSError(const util::Status& status) {
   return status.ToNSError();
 }
 
-inline Status MakeStatus(NSError* error) {
-  if (!error) {
-    return Status::OK();
-  }
+Status MakeStatus(NSError* error);
+
+using VoidErrorBlock = void (^)(NSError* _Nullable error);
+
+util::StatusCallback MakeCallback(VoidErrorBlock _Nullable block);
 
-  HARD_ASSERT(error.domain == FIRFirestoreErrorDomain,
-              "Can only translate a Firestore error to a status");
-  auto error_code = static_cast<int>(error.code);
-  HARD_ASSERT(error_code >= FirestoreErrorCode::Cancelled &&
-                  error_code <= FirestoreErrorCode::Unauthenticated,
-              "Unknown error code");
-  return Status{static_cast<FirestoreErrorCode>(error_code),
-                MakeString(error.localizedDescription)};
+template <typename T>
```

Comment:
It's impressive how a `using` declaration can work with an Objective-C block.

Reply:
Oh, interesting, I wasn't aware of this proposal. I couldn't find any updates since 2009, though. Also, I'm skeptical about its chances of going through WG21, because of the overlap with lambdas. However, adding those to C could be interesting.